src/v2/kuadrant/pipeline/tasks/ratelimit.rs

```diff
@@ -1,36 +1,785 @@
+use crate::envoy::rate_limit_descriptor::Entry;
+use crate::envoy::{rate_limit_descriptor, rate_limit_response, RateLimitDescriptor};
+use crate::v2::data::attribute::{AttributeError, AttributeState};
+use crate::v2::data::cel::{Expression, Predicate};
+use crate::v2::kuadrant::pipeline::blueprint::{Action, ConditionalData};
 #[allow(dead_code)]
-use crate::v2::data::cel::Predicate;
-use crate::v2::kuadrant::pipeline::tasks::{Task, TaskOutcome};
+use crate::v2::kuadrant::pipeline::tasks::{PendingTask, Task, TaskOutcome};
 use crate::v2::kuadrant::ReqRespCtx;
+use crate::v2::services::rate_limit::RateLimitService;
 use crate::v2::services::Service;
+use cel_interpreter::Value;
 use std::rc::Rc;
 
-#[allow(dead_code)]
-struct RateLimitTask {
+/// Builds individual descriptor entries from CEL expressions
+struct DescriptorEntryBuilder {
+    key: String,
+    expression: Expression,
+}
+
+impl DescriptorEntryBuilder {
+    /// Create an instance of DescriptorEntryBuilder
+    fn new(key: String, expression: Expression) -> Self {
+        Self { key, expression }
+    }
+
+    /// Evaluate the expression to create a descriptor entry
+    fn evaluate(self, ctx: &ReqRespCtx) -> Result<rate_limit_descriptor::Entry, AttributeError> {
+        match self.expression.eval(ctx) {
+            Ok(AttributeState::Available(value)) => {
+                let value_str = match value {
+                    Value::Int(n) => format!("{n}"),
+                    Value::UInt(n) => format!("{n}"),
+                    Value::Float(n) => format!("{n}"),
+                    Value::String(s) => s.to_string(),
+                    Value::Bool(b) => format!("{b}"),
+                    Value::Null => "null".to_owned(),
+                    _ => {
+                        return Err(AttributeError::Parse(
+                            "Only scalar values can be sent as data".to_string(),
+                        ));
+                    }
+                };
+
+                Ok(rate_limit_descriptor::Entry {
+                    key: self.key.clone(),
+                    value: value_str,
+                })
+            }
+            Ok(AttributeState::Pending) => Err(AttributeError::NotAvailable(format!(
+                "Expression '{}' evaluation is pending",
+                self.key
+            ))),
+            Err(cel_err) => Err(AttributeError::Retrieval(format!(
+                "CEL evaluation error for '{}': {}",
+                self.key, cel_err
+            ))),
+        }
+    }
+}
+
+/// Type indicating the state of the descriptors build outcome
+enum BuildDescriptorsState {
+    Ready(Vec<RateLimitDescriptor>),
+    Pending,
+}
```

Comment:
I've not reasoned out the behaviour of all of this in my head but this slightly concerns me; I _thought_ this would be handled by the current changes in the CEL library and the existing caching logic.

If it requires data at a phase it's not currently in, just 'trying' to build the descriptor is enough for it to cache things; if it then builds the descriptor later in the correct phase all of the data it needs should be there.